# Obsidian 助手 v2.0 + Token 计数测试

**目标**：测试 Obsidian 助手功能并实时监控 Token 消耗

**版本**: v2.0  
**日期**: 2025-11-13

## 1. 环境准备

In [1]:
import os
import sys
from pathlib import Path
from dotenv import load_dotenv

# 添加 deepagents 路径
sys.path.insert(0, str(Path.cwd() / "deepagents_official" / "libs"))

# 加载环境变量
load_dotenv()

# 验证 API Keys
print("✅ 环境变量检查:")
print(f"DASHSCOPE_API_KEY: {'已设置' if os.getenv('DASHSCOPE_API_KEY') else '❌ 未设置'}")
print(f"TAVILY_API_KEY: {'已设置' if os.getenv('TAVILY_API_KEY') else '❌ 未设置'}")

✅ 环境变量检查:
DASHSCOPE_API_KEY: 已设置
TAVILY_API_KEY: 已设置


## 2. 加载模块

In [2]:
# 加载 Obsidian 助手模块
exec(open('obsidian_assistant.py').read())
print("✅ Obsidian 助手 v2.0 模块已加载\n")

# 加载 Token 计数器
exec(open('token_counter.py').read())

# 加载展示工具
from utils import display_agent_execution, extract_final_answer

✅ Obsidian 助手 v2.0 模块已加载

✅ Token 计数器模块加载完成


## 3. 创建助手和计数器

In [3]:
# 创建 Obsidian 助手
assistant = create_obsidian_assistant_v2()

# 创建 Token 计数器
token_counter = TokenCounter(model="qwen-turbo")

print("\n✅ 准备就绪！开始测试...")

🔧 正在创建 Obsidian 助手 v2.0...
  ✓ 初始化模型: qwen-turbo
  ✓ 创建搜索工具（文档路径: /Users/yf/Documents/Obsidian Vault/我的知识库/Obsidian_Knowledge/obsidian-help-master）
  ✓ 配置网页搜索子代理
  ✓ 组装主代理...
✅ Obsidian 助手 v2.0 创建成功！

📋 配置信息：
  - 模型: ChatTongyi (qwen-turbo)
  - 搜索工具: search_obsidian_docs_v2（支持路径返回）
  - 子代理: web-search-agent-v2（增强触发逻辑）
  - 引用格式: Obsidian 内部链接 + 网页链接
  - 文档路径: /Users/yf/Documents/Obsidian Vault/我的知识库/Obsidian_Knowledge/obsidian-help-master


✅ 准备就绪！开始测试...
  ✓ 创建搜索工具（文档路径: /Users/yf/Documents/Obsidian Vault/我的知识库/Obsidian_Knowledge/obsidian-help-master）
  ✓ 配置网页搜索子代理
  ✓ 组装主代理...
✅ Obsidian 助手 v2.0 创建成功！

📋 配置信息：
  - 模型: ChatTongyi (qwen-turbo)
  - 搜索工具: search_obsidian_docs_v2（支持路径返回）
  - 子代理: web-search-agent-v2（增强触发逻辑）
  - 引用格式: Obsidian 内部链接 + 网页链接
  - 文档路径: /Users/yf/Documents/Obsidian Vault/我的知识库/Obsidian_Knowledge/obsidian-help-master


✅ 准备就绪！开始测试...


## 4. 测试 1：本地知识库搜索 + Token 统计

In [ ]:
question_1 = "如何在 Obsidian 中创建双向链接？"

print("🤔 问题:", question_1)
print("=" * 80)

# 开始计时
token_counter.start_counting()

# 调用助手
result_1 = assistant.invoke({"messages": [("user", question_1)]})
display_agent_execution(result_1)

# 统计 Token
record_1 = count_tokens_for_result(question_1, result_1, token_counter)
token_counter.print_current_usage(record_1)

🤔 问题: 如何在 Obsidian 中创建双向链接？
🤖 DeepAgents 执行流程

📝 步骤 1: 用户输入
────────────────────────────────────────────────────────────────────────────────
💬 如何在 Obsidian 中创建双向链接？

🤔 步骤 2: Agent 响应
────────────────────────────────────────────────────────────────────────────────
💡 Agent 回复:
  在 Obsidian 中创建双向链接是一种非常有用的功能，它允许你在不同的笔记之间建立联系。以下是创建双向链接的步骤：

1. **内部链接**：使用双方括号 `[[ ]]` 包裹笔记名称。例如，如果你有一个名为 "Example" 的笔记，你可以写成 `[[Example]]`（参考：[[Linking notes and files/Internal links|内部链接]]）。

2. **外部链接**：使用 Markdown 格式 `[文本](URL)`。例如，如果你要链接到一个网页，可以写成 `[示例网站](https://example.com)`（参考：[[Editing and formatting/Basic formatting syntax|基础格式化语法]]）。

3. **双向链接**：当你在笔记中创建一个内部链接时，Obsidian 会自动为你创建一个反向链接。这意味着如果你在 "Note A" 中链接到 "Note B"，那么在 "Note B" 中也会出现一个链接指向 "Note A"（参考：[[Linking notes and files/Bi-directional links|双向链接]]）。

### 参考来源
⭐⭐⭐ [[Linking notes and files/Internal links|内部链接]] - 核心参考
⭐⭐ [[Editing and formatting/Basic formatting syntax|基础格式化语法]] - 补充阅读
⭐⭐ [[Linking notes and files/Bi-directional links|双向链接]] - 

## 5. 测试 2：网页搜索触发 + Token 统计

In [ ]:
question_2 = "推荐一些最新的 Obsidian 效率插件"

print("\n🤔 问题:", question_2)
print("=" * 80)

# 开始计时
token_counter.start_counting()

# 调用助手
result_2 = assistant.invoke({"messages": [("user", question_2)]})
display_agent_execution(result_2)

# 统计 Token
record_2 = count_tokens_for_result(question_2, result_2, token_counter)
token_counter.print_current_usage(record_2)


🤔 问题: 推荐一些最新的 Obsidian 效率插件
🤖 DeepAgents 执行流程

📝 步骤 1: 用户输入
────────────────────────────────────────────────────────────────────────────────
💬 推荐一些最新的 Obsidian 效率插件

🤔 步骤 2: Agent 响应
────────────────────────────────────────────────────────────────────────────────
🛠️  Agent 决定调用工具:
  1. 工具名称: task
     参数: {
      "description": "搜索最新的 Obsidian 效率插件",
      "subagent_type": "web-search-agent-v2"
}
💡 Agent 回复:
  为了推荐一些最新的 Obsidian 效率插件，我将使用网络搜索代理来查找相关信息。



⚙️  步骤 3: 工具执行结果
────────────────────────────────────────────────────────────────────────────────
🔧 工具: task
📊 返回结果:
  根据搜索结果，以下是一些关于 Obsidian 最新效率插件的信息：

1. **[The Outer Worlds 2's best feature is the one they never added - Polygon](https://www.polygon.com/the-outer-worlds-2-best-feature/)**  
   虽然这篇帖子主要讨论的是《The Outer Worlds 2》游戏的特性，但提到了 Obsidian 的开发理念，可能与 Obsidian 效率插件的设计思想有关。  
   **可信度：第三方**

2. **[What Didn’t Make It Into The Outer Worlds 2 - Unlocked Clips - IGN India](https://in.ign.com/outer-worlds-2/246584/video/what-didnt-

## 6. 测试 3：混合搜索 + Token 统计

In [ ]:
question_3 = "Canvas 功能是什么？有没有最新的使用教程？"

print("\n🤔 问题:", question_3)
print("=" * 80)

# 开始计时
token_counter.start_counting()

# 调用助手
result_3 = assistant.invoke({"messages": [("user", question_3)]})
display_agent_execution(result_3)

# 统计 Token
record_3 = count_tokens_for_result(question_3, result_3, token_counter)
token_counter.print_current_usage(record_3)


🤔 问题: Canvas 功能是什么？有没有最新的使用教程？
🤖 DeepAgents 执行流程

📝 步骤 1: 用户输入
────────────────────────────────────────────────────────────────────────────────
💬 Canvas 功能是什么？有没有最新的使用教程？

🤔 步骤 2: Agent 响应
────────────────────────────────────────────────────────────────────────────────
💡 Agent 回复:
  Canvas 是 Obsidian 中的一个功能，它允许用户在笔记中创建和管理可视化内容，如流程图、思维导图、图表等。通过 Canvas，用户可以将抽象的想法和概念以图形化的方式呈现出来，从而更好地组织和理解信息。

关于最新的使用教程，您可以参考以下资源：

1. **Obsidian 官方文档**：提供了关于 Canvas 功能的详细说明和使用指南（参考：[Obsidian 官方文档](https://obsidian.md)）。
2. **Obsidian 社区论坛**：社区成员经常分享他们的经验和教程，您可以在这里找到一些实用的示例和技巧（参考：[Obsidian 社区论坛](https://discord.gg/6U4ZmW7)）。
3. **YouTube 教程**：许多 Obsidian 用户在 YouTube 上发布了关于 Canvas 功能的视频教程，这些视频通常会更直观地展示如何使用 Canvas。

如果您需要更具体的指导或帮助，请随时告诉我！ 

### 参考来源
- [Obsidian 官方文档](https://obsidian.md)
- [Obsidian 社区论坛](https://discord.gg/6U4ZmW7)

✅ 执行完成


📊 Token 使用统计
🕐 时间: 2025-11-13 22:42:46
❓ 问题: Canvas 功能是什么？有没有最新的使用教程？
⏱️  响应时间: 3.55 秒

💬 Token 详情:
  - 输入:      29
  - 输出:     486
  - 总计:     515

💰 预估成本: ¥0.0030 元




## 7. 累积统计报告

In [7]:
# 打印完整统计报告
token_counter.print_statistics()


📈 累积统计报告
📞 总调用次数: 3
⏱️  总响应时间: 22.07 秒
⚡ 平均响应: 7.36 秒/次

💬 Token 统计:
  - 累计 Token:      2,259
  - 平均 Token:        753

💰 成本统计:
  - 累计成本: ¥0.0133 元
  - 平均成本: ¥0.0044 元/次



## 8. 结果分析

### 📊 预期观察

1. **本地搜索** (测试 1)
   - 响应时间: ~3-5秒
   - Token 消耗: ~2,000-4,000
   - 成本: ~¥0.01

2. **网页搜索** (测试 2)
   - 响应时间: ~30-40秒 ⚠️
   - Token 消耗: ~8,000-15,000 ⚠️
   - 成本: ~¥0.05 ⚠️

3. **混合搜索** (测试 3)
   - 响应时间: ~10-20秒
   - Token 消耗: ~5,000-10,000
   - 成本: ~¥0.02-0.03

### 📊 测试结果分析

1️⃣ 性能对比
- 纯本地搜索: 4.31秒，512 tokens，¥0.0030
- 网页搜索: 14.22秒，1,232 tokens，¥0.0073（2.4倍 tokens，2.4倍成本）
- 混合搜索: 3.55秒，515 tokens，¥0.0030

2️⃣ 关键发现
- ⚡ 本地检索非常高效：3-4秒内完成，tokens 保持在 ~500
- 🌐 网页搜索开销大：14秒响应，1,232 tokens（增加 140%）
- 🎯 混合模式最优：优先本地，仅在需要时补充网页

3️⃣ 成本预估（按每天100次查询）
- 纯本地：¥0.30/天 = ¥9/月
- 纯网页：¥0.73/天 = ¥22/月
- 混合模式（70%本地+30%网页）：¥0.43/天 = ¥13/月

### 🎯 优化方向

根据测试结果，考虑：
- ✅ 添加缓存（减少重复搜索）
- ✅ 实现多模式（经济/平衡/专业）
- ✅ 智能路由（自动选择策略）